### DISCUSSION TITLE
Java: How to know if a thrown exception is unnecessary?

### DISCUSSION THREAD

**Discussion by negora** [apache/netbeans#4086] [Q&A]
Hi:

Playing around with other IDEs, I've seen that both Eclipse IDE and IntelliJ IDEA are able to show if a checked exception is unnecessary in a `throws` clause. Is there a similar function in NetBeans? By *unnecessary* I mean that no line in the method's body throws such exception.

In all years I've been using NetBeans (17) I never thought about it, sincerely. But that feature has been really useful to remove exceptions (from `throws` clauses) that I left by mistake after big code refactorings.

By the way, I want to clarify that Eclipse detects these unnecessary exceptions live, in any kind of method, be it private or non-private. IntelliJ, on his part, detects them live in private methods only. For non-private methods, you need to run a code analysis. To me, the behavior of IntelliJ would be enough.

Thank you.

**Comment by mbien** [comment]
Hi @negora,

well, there is more or less :) If you put the text cursor over the Exception (works for both, catch clauses and throws clauses), NetBeans will highlight the lines of code which can throw it. If there is nothing highlighted it should be safe to remove (unless its public).

I checked and I don't think that there is a code inspection for that, unless I overlooked something. Could be potentially added I suppose since the functionally is basically already there in some form.

**Reply by negora** [reply]
@mbien

> well, there is more or less :) If you put the text cursor over the Exception (works for both, catch clauses and throws clauses), NetBeans will highlight the lines of code which can throw it. If there is nothing highlighted it should be safe to remove (unless its public).

From time to time I use that feature while I'm editing a method, but it's very time consuming to check every method that way. But thank you for the proposal, of course ;-) .

Some kind of massive inspection would be enough in my opinion. No need to make this inspection in real time.

> Could be potentially added I suppose since the functionally is basically already there in some form.

That's what I thought, that the functionally is someway already there. But I don't know if there is some kind of technical limitation. Because there are some hints, such that one that checks if the package of a class matches the directory tree, that are not available in the *Inspection and transform* menu. I don't know why.

**Comment by negora** [comment]
About this discussion, Can it be converted into a proper issue? (a feature request in this case)

Thank you!
